File: v6/lib/command_data.dart
import 'package:bad_word_blocker_dart/globals.dart';
import 'package:nyxx/nyxx.dart';
import 'package:nyxx_interactions/nyxx_interactions.dart';
import 'events/slash_interaction.dart';

void registerCommands(
    INyxxWebsocket client, IInteractions interaction, bool upload) {
  final pingCommand = SlashCommandBuilder("ping", "Pings the bot", [])
    ..registerHandler(onSlashCommandInteraction);

  final blacklistCommand = SlashCommandBuilder(
      "blacklist", "Modify this server's blacklist", [],
      requiredPermissions: PermissionsConstants.manageMessages)
    ..registerHandler(onSlashCommandInteraction);

  final bypassCommand = SlashCommandBuilder(
      "bypass",
      "Manage bypasses",
      [
        CommandOptionBuilder(CommandOptionType.subCommand, "role",
            "Add a role that will bypass the bot",
            options: [
              CommandOptionBuilder(CommandOptionType.role, "role",
                  "Role to bypass. Anyone with this role will bypass the bot's filter.",
                  required: true)
            ])
          ..registerHandler(onSlashCommandInteraction),
        CommandOptionBuilder(CommandOptionType.subCommand, "channel",
            "Add a channel that will bypass the bot",
            options: [
              CommandOptionBuilder(CommandOptionType.channel, "channel",
                  "Channel to ignore. Every message sent in this channel will bypass the bot's filter.",
                  required: true, channelTypes: allowedCommandChannelTypes)
            ])
          ..registerHandler(onSlashCommandInteraction),
        CommandOptionBuilder(
          CommandOptionType.subCommand,
          "manage",
          "View/Remove current bypassing channels and roles",
        )..registerHandler(onSlashCommandInteraction)
      ],
      requiredPermissions: PermissionsConstants.manageMessages);

  final customizeEmbedCommand = SlashCommandBuilder("customize_embed",
      "Customize the embed that's sent when a message is blocked", [],
      requiredPermissions: PermissionsConstants.manageMessages)
    ..registerHandler(onSlashCommandInteraction);

  final strikesCommand = SlashCommandBuilder(
      "strikes",
      "Manage strikes",
      [
        CommandOptionBuilder(
          CommandOptionType.subCommand,
          "view",
          "View a person's strikes",
          options: [
            CommandOptionBuilder(CommandOptionType.user, "user",
                "User whose strikes you want to see",
                required: true)
          ],
        )..registerHandler(onSlashCommandInteraction),
        CommandOptionBuilder(CommandOptionType.subCommand, "edit",
            "Change someone's strikes to a desired amount",
            options: [
              CommandOptionBuilder(
                  CommandOptionType.user, "user", "User whose strike to change",
                  required: true),
              CommandOptionBuilder(CommandOptionType.number, "amount",
                  "Amount to change their strikes to",
                  required: true)
            ])
          ..registerHandler(onSlashCommandInteraction)
      ],
      requiredPermissions: PermissionsConstants.manageMessages);

  final limitsCommand = SlashCommandBuilder(
      "limits",
      "Manage limits",
      [
        CommandOptionBuilder(
            CommandOptionType.subCommand, "add", "Add a limit on strikes",
            options: [
              CommandOptionBuilder(CommandOptionType.integer, "amount",
                  "Amount of strikes that should trigger this limit",
                  required: true),
              CommandOptionBuilder(CommandOptionType.string, "action",
                  "What should be done when this limit is triggered",
                  choices: [
                    ArgChoiceBuilder("ban", "ban"),
                    ArgChoiceBuilder("kick", "kick"),
                    ArgChoiceBuilder("timeout", "timeout"),
                    ArgChoiceBuilder("assign mute role", "mute_role")
                  ],
                  required: true),
              CommandOptionBuilder(CommandOptionType.integer, "duration",
                  "If desired, how many minutes this action should last")
            ])
          ..registerHandler(onSlashCommandInteraction),
        CommandOptionBuilder(
            CommandOptionType.subCommand, "manage", "View/remove limits")
          ..registerHandler(onSlashCommandInteraction)
      ],
      requiredPermissions: PermissionsConstants.manageMessages);

  final logCommand = SlashCommandBuilder(
      "log",
      "Manages log channel",
      [
        CommandOptionBuilder(CommandOptionType.subCommand, "set",
            "Set channel that will receive updates when a message is blocked",
            options: [
              CommandOptionBuilder(CommandOptionType.channel, "channel",
                  "The channel you want the bot to sent messages in when a message is blocked",
                  required: true, channelTypes: allowedCommandChannelTypes)
            ])
          ..registerHandler(onSlashCommandInteraction),
        CommandOptionBuilder(CommandOptionType.subCommand, "view",
            "View the current log channel")
          ..registerHandler(onSlashCommandInteraction),
        CommandOptionBuilder(CommandOptionType.subCommand, "remove",
            "Remove the current log channel")
          ..registerHandler(onSlashCommandInteraction)
      ],
      requiredPermissions: PermissionsConstants.manageMessages);

  final getCommand = SlashCommandBuilder("get",
      "Get back your last deleted message, along with why it was deleted", [])
    ..registerHandler(onSlashCommandInteraction);

  final helpCommand =
      SlashCommandBuilder("help", "Get help with Bad Word Blocker", [])
        ..registerHandler(onSlashCommandInteraction);

  final serverCCommand = SlashCommandBuilder(
      "serverc", "See how many servers Bad Word Blocker is in", [])
    ..registerHandler(onSlashCommandInteraction);

  final premiumCommand = SlashCommandBuilder(
      "premium",
      "Manages premium aspects of the bot",
      [
        CommandOptionBuilder(
          CommandOptionType.subCommand,
          "manage",
          "Manage your premium status and data",
        )..registerHandler(onSlashCommandInteraction),
        CommandOptionBuilder(CommandOptionType.subCommand, "add",
            "Add this server to your premium subscription")
          ..registerHandler(onSlashCommandInteraction),
      ],
      requiredPermissions: PermissionsConstants.manageMessages);

  final i = interaction
    ..registerSlashCommand(pingCommand)
    ..registerSlashCommand(blacklistCommand)
    ..registerSlashCommand(bypassCommand)
    ..registerSlashCommand(customizeEmbedCommand)
    ..registerSlashCommand(strikesCommand)
    ..registerSlashCommand(limitsCommand)
    ..registerSlashCommand(logCommand)
    ..registerSlashCommand(getCommand)
    ..registerSlashCommand(helpCommand)
    ..registerSlashCommand(serverCCommand)
    ..registerSlashCommand(premiumCommand);
  if (upload) i.syncOnReady();
}

File: v5/CommandData.fs
module CommandData

open Discord
open Types
open GlobalVariables
open Wrappers.DiscordWrappers

let (commands: ApplicationCommandProperties array) =
    [| { Name = "ping"
         Description = "Ping the bot"
         Options = [||]
         DefaultPermission = None }
       { Name = "blacklist"
         Description = "Edit the server's blacklist"
         Options = [||]
         DefaultPermission = Some GuildPermission.ManageMessages }
       { Name = "bypass"
         Description = "Parent bypass command"
         DefaultPermission = Some GuildPermission.ManageMessages
         Options =
           [| { OptionType = SubCommand
                Name = "channel"
                Description = "Ignores a channel"
                Required = false
                Options =
                  [| { OptionType =
                         SlashCommandOptionType.Channel
                             [ ChannelType.Text
                               ChannelType.PublicThread
                               ChannelType.PrivateThread
                               ChannelType.News
                               ChannelType.NewsThread
                               ChannelType.Forum ]
                       Name = "channel"
                       Description = "Channel that should be bypassed"
                       Required = true
                       Options = [||] } |] }
              { OptionType = SubCommand
                Name = "role"
                Description = "Add an ignored role"
                Required = false
                Options =
                  [| { OptionType = Role
                       Name = "role"
                       Description = "Role to add. Anyone with this role will be ignored by the bot"
                       Required = true
                       Options = [||] } |] }
              { OptionType = SubCommand
                Name = "manage"
                Description = "View and remove bypassing channels and roles"
                Required = false
                Options = [||] }

              |] }
       { Name = "strikes"
         Description = "Strikes parent command"
         DefaultPermission = Some GuildPermission.ManageMessages
         Options =
           [| { OptionType = SubCommand
                Name = "view"
                Description = "View a person's strikes"
                Required = false
                Options =
                  [| { OptionType = User
                       Name = "user"
                       Description = "User whose strikes you'd like to view"
                       Required = false
                       Options = [||] } |] }
              { OptionType = SubCommand
                Name = "edit"
                Description = "Sets a person's strikes"
                Required = false
                Options =
                  [| { OptionType = User
                       Name = "user"
                       Description = "User whose strikes you'd like to change"
                       Required = true
                       Options = [||] }
                     { OptionType = Number <| Some(0, 250)
                       Name = "amount"
                       Description = "Amount of strikes they should have"
                       Required = true
                       Options = [||] } |] } |] }

       { Name = "limits"
         Description = "Limits parent command"
         DefaultPermission = Some GuildPermission.ManageMessages
         Options =
           [| { OptionType = SubCommand
                Name = "add"
                Description = "Add a new limit on strikes"
                Required = false
                Options =
                  [| { OptionType = Number <| Some(1, 250)
                       Name = "amount"
                       Description = "Amount of strikes that should trigger this limit"
                       Required = true
                       Options = [||] }
                     { OptionType =
                         String
                         <| Some
                             [| ("Ban", "ban")
                                ("Kick", "kick")
                                ("Timeout", "timeout")
                                ("Use mute role", "role") |]
                       Name = "action"
                       Description = "What should be done to the user when they reach the amount of strikes"
                       Required = true
                       Options = [||] }
                     { OptionType = Number <| Some(1, jsonInfo.banLimit)
                       Name = "minutes"
                       Description = "How long the action should last"
                       Required = false
                       Options = [||] } |] }
              { OptionType = SubCommand
                Name = "manage"
                Description = "View and remove limits"
                Required = false
                Options = [||] } |] }
       { Name = "custom_embed"
         Description = "Customizes the embed that is sent when a message is blocked"
         Options = [||]
         DefaultPermission = Some GuildPermission.ManageMessages }
       { Name = "log"
         Description = "Parent log command"
         DefaultPermission = Some GuildPermission.ManageMessages
         Options =
           [| { OptionType = SubCommand
                Name = "set"
                Description = "Sets a channel that will be notified when a message is blocked"
                Required = false
                Options =
                  [| { OptionType =
                         SlashCommandOptionType.Channel
                             [ ChannelType.Text; ChannelType.PrivateThread; ChannelType.PublicThread ]
                       Name = "channel"
                       Description = "Channel that should be notified when a message is blocked"
                       Required = true
                       Options = [||] } |] }
              { OptionType = SubCommand
                Name = "manage"
                Description = "View log channel and optionally removes it via button"
                Required = false
                Options = [||] } |] }
       { Name = "get"
         Description = "Gets your last message back"
         Options = [||]
         DefaultPermission = None }
       { Name = "serverc"
         Description = "See how many servers Bad Word Blocker is in"
         Options = [||]
         DefaultPermission = None }
       { Name = "extras"
         Description = "Enables extra features that are unrelated to message filtering"
         Options = [||]
         DefaultPermission = Some GuildPermission.ManageMessages } |]

    |> Array.map createSlashCommand
    |> Array.map (fun x -> x :> ApplicationCommandProperties)

let (extrasCommands: ApplicationCommandProperties array) =
    [| { Name = "commchannel"
         Description = "Command channel parent command"
         DefaultPermission = Some GuildPermission.ManageMessages
         Options =
           [| { OptionType = SubCommand
                Name = "add"
                Description =
                  "(extras) Adds a channel where, upon a message being sent, it's deleted unless it's from a bot"
                Required = false
                Options =
                  [| { OptionType = SlashCommandOptionType.Channel [ ChannelType.Text ]
                       Name = "channel"
                       Description = "Channel to add to list of command channels"
                       Required = true
                       Options = [||] } |] }
              { OptionType = SubCommand
                Name = "manage"
                Description = "(extras) View and remove command channels"
                Required = false
                Options = [||] } |] }
       { Name = "cleanupbot"
         Description = "Cleanupbot parent command"
         DefaultPermission = Some GuildPermission.ManageMessages
         Options =
           [| { OptionType = SubCommand
                Name = "add"
                Description = "(extras) Add a bot whose messages will be deleted after a given amount of seconds"
                Required = false
                Options =
                  [| { OptionType = User
                       Name = "bot"
                       Description = "Bot whose messages should be deleted after the given seconds"
                       Required = true
                       Options = [||] }

                     { OptionType = Number <| Some(3, 1000)
                       Name = "seconds"
                       Description = "Amount of seconds that the bot's message should be deleted after"
                       Required = true
                       Options = [||] } |] }
              { OptionType = SubCommand
                Name = "manage"
                Description = "(extras) View and remove cleanup bots"
                Required = false
                Options = [||] } |] } |]
    |> Array.map createSlashCommand
    |> Array.map (fun x -> x :> ApplicationCommandProperties)
